### PR TITLE
Factor out the DNS-over-UDP code

### PR DESCRIPTION
**Pull request by erikcorry**
This mostly just moved the DNS-over-UDP code to its own lib file.

The lookup function takes a block which creates the right sort of
UDP socket if needed.

The idea of this is to make it easier to use the DNS-over-UDP code
for cellular drivers.

### CODE REVIEW

File: lib/dns_over_udp.toit [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+  dots := 0
+  name.do:
+    if it == '.':
+      dots++
+    else:
+      if not '0' <= it <= '9': return false
+  return dots == 3
+
+/**
+Look up a domain name.
```
Comment (lask) [inline lib/dns_over_udp.toit:33]:
```suggestion
Looks up a domain name.
```

File: lib/dns_over_udp.toit [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    --timeout/Duration=DNS_DEFAULT_TIMEOUT:
+  if ip_string_ host:
+    return net.IpAddress.parse host
+  q := DnsQuery_ host
+  return q.get --server=server --timeout=timeout get_udp_socket
+
+class DnsQuery_:
+  static A     ::= 1  // A address.
+  static CNAME ::= 5  // Canonical name.
+  static CLASS ::= 1  // The Internet class.
```
Comment (lask) [inline lib/dns_over_udp.toit:59]: IMO make the names so the comments are unnecessary by including all the necessary information, e.g. ` A_ADDRESS`.

File: lib/dns_over_udp.toit [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+      --timeout/Duration=DNS_DEFAULT_TIMEOUT
+      [get_udp_socket]:
+    if HOSTS_.contains name:
+      return net.IpAddress.parse HOSTS_[name]
+
+    hit := find_in_cache_ server
+    if hit: return hit
+
+    query := create_query_
+    socket := get_udp_socket.call
```
Comment (lask) [inline lib/dns_over_udp.toit:88]: Maybe add a type to have a dynamic check that it is the correct socket.
```suggestion
    socket/udp.Socket := get_udp_socket.call
```
You would also need to have the import:
```
import net.udp
```
Finally, we should probably mention it in the toit doc.

Comment (lask) [reply lib/dns_over_udp.toit:88]: I looked at the previous implementation and it should create a socket of the correct type.

File: lib/dns_over_udp.toit [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+        retry_timeout := DNS_RETRY_TIMEOUT
+
+        // Resend the query with exponential backoff until the outer timeout
+        // expires.
+        while true:
+          socket.write query
+
+          answer := null
+          exception := catch:
```
Comment (lask) [inline lib/dns_over_udp.toit:104]:
```suggestion
          exception := catch --trace=(: it != "DEADLINE_EXCEEDED"):
```

File: lib/dns_over_udp.toit [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+            with_timeout retry_timeout:
+              answer = socket.receive
+          if exception and exception != "DEADLINE_EXCEEDED": throw exception
+
+          if answer:
+            return decode_response_ answer.data server
```
Comment (lask) [inline lib/dns_over_udp.toit:110]: One-liner like above?

**Review by lask — COMMENTED** [review]
I think this is a great candidate for a package, but I am also fine with it as a library.

It is less inconvenient for user when we update the interface of a a package than when we update  the interface of a library.

**Comment by erikcorry** [comment]
This is probably not the approach we will go with.

**Pull request closed without merging**